### ISSUE TITLE
Query for specific fields

### ISSUE THREAD

**Issue by davidlesieur** [datastore/datastore#11]
I'm exploring the library; good work!

One feature I find missing is the ability for a `Query` to specify fields to retrieve, instead of retrieving full objects. MongoDB allows this. I'm not too familiar with other datastores, but I guess a query could always return full objects if the datastore does not support selecting fields.

What would you think about such feature?

Thanks!

BTW, issue tracking is not enabled on https://github.com/datastore/datastore. Not sure if that is intentional or not, so I thought I'd let you know.

**Comment by jbenet** [comment]
> I'm exploring the library; good work!

Thanks! :)  Glad you like it. All feedback very welcome!

> One feature I find missing is the ability for a Query to specify fields to retrieve

Hm, yeah, retrieving specific fields could be useful. How does this play nicely with the datastore abstraction that a data value is not touched?

It may be the case that the more complicated queries (filters/orders/etc) and this field selecting could be abstracted into a plugin (`datastore.query`) or something, that extends basic query translators?

There is a fine line to walk between not breaking the KV-store abstraction in the whole library, and fully supporting relational databases. Most of my (and @athenalabs)  `datastore` use is actually as a pure KV-store (redis, filesystems, s3). 

> BTW, issue tracking is not enabled on https://github.com/datastore/datastore.

Thanks -- i need to transfer this repo (jbenet/datastore) into the @datastore org, and then link `jbenet/datastore --> datastore/datastore`). Just not sure whether doing that will break anyone elses pointers -- haven't gotten around to investigating that and moving it.

**Comment by davidlesieur** [comment]
> It may be the case that the more complicated queries (filters/orders/etc) and this field selecting could be abstracted into a plugin (datastore.query) or something, that extends basic query translators?

Yes, I guess it could be done in some extension. However, the way I saw this the abstraction was not being broken, as selecting fields would just be a feature of `datastore.query`, and implementations would be free to ignore that feature (and return full objects). It would be merely an optimization option for callers who do not need all the fields from the objects.

It would be up to the developer to choose the right tool for the job and, as for the filtering and ordering features, to choose the right implementation if field selection is important.

**Comment by jbenet** [comment]
Additional thoughts (thinking out loud):

Another benefit for keeping it inside the main library:
- It can become tedious and problematic to have plugins depend on each other (e.g. `datastore.mongo` depends on `datastore.query` depends on `datastore`). 

Another benefit of moving it out:
- would resolve the question of whether query should or should-not be implemented by all datastores. Default is no, and you can implement it optionally. There can be a simple wrapper datastore that adds the basic query functionality.

Hm, this question boils down to whether or not acknowledging "fields" in object data is a good thing for the main library to do. It's certainly not at the KVS-generic level; it assumes a traditional Relational structure. Even "document" databases are relational-ish. 

I do think the ability to retrieve "lists" of objects is crucial though... It's very annoying to have hash-tables without being able to list them. So `query` is nice for this: 

```
# retrieve all objects under `/foo`:
datastore.query(Query(Key(`/foo`)))

# aside, this should be the same as:
datastore.query('/foo')
datastore.query(Query('/foo'))
```

Requiring all `datastore` implementations to support this may be a plus. e.g. `datastore.memcache` _could_ wrap its base datastore with a CollectionDatastore. However, this again may be a thing to leave up to the user.
